### PR TITLE
Prayers Again

### PR DESCRIPTION
**Pull request by Just-a-Unity-Dev**
## About the PR
So let's say you've wanted to do a gimmick you had in mind for a few rounds, instead of ahelping and possibly bothering admin's doing work, go to your local chapel and pray to the gods! Nanotrasen Inc. is not responsible for injury, death or catgirl-ification.

Admins can also subtly message users instead of bothering them in the ahelp menu, supports customized text (e.g "Your radio crackles to life" or leave it blank to default to "You hear a voice in your head...")

**Screenshots**
image
image
image
(exploding smite isn't part of this PR but it fits in)
image

**Changelog**
:cl: eclips_e
- add: Nanotrasen's R&D department has harnessed the power of talking to the god's, essentially, you can pray at altars now! Nanotrasen Inc. is not responsible for injury, death or catgirl-ification.
- add: Added the ability for chaplain's to pray at their bible.

### CODE REVIEW

**Comment by Visne** [comment]
Should be another chat mode IMO, instead of a "command" which we don't use anywhere else as far as I know and is hidden from the user.

**Comment by Just-a-Unity-Dev** [comment]
> Should be another chat mode IMO, instead of a "command" which we don't use anywhere else as far as I know and is hidden from the user.

commands can be used in the chat wym? unless i'm misreading your sentence

**Comment by EmoGarbage404** [comment]
> Should be another chat mode IMO, instead of a "command" which we don't use anywhere else as far as I know and is hidden from the user.

i mean, i understand what you mean, but having a new chat mode for a single command seems kinda silly. Also, modifying chat modes/adding new ones is likely to conflict with hud refactor. Perhaps just make this a text box that pops up when you interact with an altar, as those are present in most maps. Would also give chaplains something fun.

**Comment by mirrorcult** [comment]
honestly i kind of want praying to exist but i dont want it to be as accessible as a full chat mode

**Comment by Just-a-Unity-Dev** [comment]
the PR will be up for review tomorrow

**Comment by Visne** [comment]
Altar seems fine to me. 

In my opinion it doesn't have to be a chat mode per se, but it should be if the alternative is a `/pray` command (which is hidden to the player).

**Comment by mirrorcult** [comment]
oh yeah wait it should 100% use the altars that's way cooler

**Comment by asperger-sind** [comment]
Altars are sick, albeit I think that you should be able to /pray under any circumstances as a ghostrole or an antag (nukies for example), as said nukies lack an altar, but you still really feel like doing that one sick gimmick you've thought about for 5 rounds now.

**Comment by mirrorcult** [comment]
nukies should have to infiltrate the chapel if they want to pray for their gimmick

**Comment by Just-a-Unity-Dev** [comment]
feeling like cutting the scope of this PR from prayers to just subtle messages, and i'll probably add prayers in a different PR

**Comment by Just-a-Unity-Dev** [comment]
hmm today I will finish a PR for the first time in a while

**Comment by Just-a-Unity-Dev** [comment]
Ready for review

File: Content.Server/Prayer/PrayerSystem.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+            SubscribeLocalEvent<PrayableComponent, GetVerbsEvent<ActivationVerb>>(AddPrayVerb);
+        }
+
+        private void AddPrayVerb(EntityUid uid, PrayableComponent comp, GetVerbsEvent<ActivationVerb> args)
+        {
+            // if it doesn't have an actor and we can't reach it then don't add the verb
+            if (!EntityManager.TryGetComponent<ActorComponent?>(args.User, out var actor))
+                return;
+            if (!args.CanAccess)
```
Comment (Veritius) [inline Content.Server/Prayer/PrayerSystem.cs:36]: Nitpick, but I would prefer an extra line between these.

File: Resources/Locale/en-US/prayers/prayers.ftl [hunk 2]
```diff
@@ -0,0 +1,7 @@
+﻿prayer-verbs-subtle-message = Subtle Message
+prayer-verbs-pray = Pray
+prayer-chat-notify = PRAYER: {$username}: {$message}
+
+prayer-popup-subtle-default = You hear a voice in your head...
+prayer-popup-notify-sent = Your message has been sent to the gods...
+prayer-popup-notify-locked = You are not worthy to pray to me.
```
Comment (Veritius) [inline Resources/Locale/en-US/prayers/prayers.ftl:7]: This seems like someone's actually talking to the person. Say something like "You don't feel worthy..." instead.

File: Content.Server/Prayer/PrayerSystem.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            prayerVerb.Impact = LogImpact.Low;
+            args.Verbs.Add(prayerVerb);
+        }
+
+        /// <summary>
+        /// Subtly messages a player by giving them a popup and a chat message.
+        /// </summary>
+        /// <param name="target"></param>
+        /// <param name="messageString"></param>
+        /// <param name="popupMessage"></param>
```
Comment (Veritius) [inline Content.Server/Prayer/PrayerSystem.cs:64]: Add descriptions to these params.

File: Content.Server/Prayer/PrayableComponent.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    /// <summary>
+    /// Allows an entity to be prayed on in the context menu
+    /// </summary>
+    [RegisterComponent]
+    public sealed class PrayableComponent : Component
+    {
+        /// <summary>
+        /// If bible users are only allowed to use this prayable entity
+        /// </summary>
+        [DataField("bibleOnly")]
```
Comment (Veritius) [inline Content.Server/Prayer/PrayableComponent.cs:12]: Ambiguous datafield name could confuse people only working with YML.

**Review by Veritius — CHANGES_REQUESTED** [review]

File: Resources/Locale/en-US/prayers/prayers.ftl (hunk 2 above)
Comment (Just-a-Unity-Dev) [reply Resources/Locale/en-US/prayers/prayers.ftl:7]: the first line is intentional because that's the admin's subtle popup

**Review by mirrorcult — CHANGES_REQUESTED** [review]
png needs to be way lower res. look at the other svg/png pairs in the folder and just emulate that

File: Content.Server/Prayer/PrayerSystem.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        /// You may be wondering, "Why the admin chat, specifically? Nobody even reads it!"
+        /// Exactly.
+        ///  </remarks>
+        public void Pray(IPlayerSession sender, string message)
+        {
+            if (sender.AttachedEntity == null)
+                return;
+
+            _popupSystem.PopupEntity(Loc.GetString("prayer-popup-notify-sent"), sender.AttachedEntity.Value, Filter.Empty().AddPlayer(sender), PopupType.Medium);
+            _chatManager.SendAdminAnnouncement(Loc.GetString("prayer-chat-notify", ("message", message), ("username", sender.Name)));
```
[the remark this answers is not in the document]
Comment (Just-a-Unity-Dev) [reply Content.Server/Prayer/PrayerSystem.cs:90]: how would i achieve this? i cant seem to find any functions in the ahelp (that or i'm blind)

**Comment by Just-a-Unity-Dev** [comment]
i'll do reviews at the end of the week because my laptop will die if i attempt to run rider

**Comment by Just-a-Unity-Dev** [comment]
Hmmmm today I will finish prayers (clueless)

image

File: Content.Server/Prayer/PrayerSystem.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Just-a-Unity-Dev) [reply Content.Server/Prayer/PrayerSystem.cs]: did it

**Comment by Just-a-Unity-Dev** [comment]
> png needs to be way lower res. look at the other svg/png pairs in the folder and just emulate that

oh right i forgot about this

**Comment by Just-a-Unity-Dev** [comment]
> png needs to be way lower res. look at the other svg/png pairs in the folder and just emulate that

@mirrorcult done

**Comment by Just-a-Unity-Dev** [comment]
why are tests failing i fucking changes a png file

**Pull request merged**